h5py/__init__.py

```diff
@@ -15,6 +15,15 @@
 from warnings import warn as _warn
 import atexit
 
+# Tell mpi4py to not automatically initialize MPI at import;
+# required to avoid problems with 'import h5py' in some environment
+# when h5py was built with MPI support
+try:
+    import mpi4py
+    mpi4py.rc.initialize = False
```

Comment:
As someone not using MPI, I'd quite like to preserve the current pattern where a non-MPI build of h5py never tries to touch mpi4py at all.

Does it work to set this setting everywhere we already import `mpi4py`, without adding an import?

Reply:
The problem is that setting the `rc.initialize` to `False` must be done as already as possible, before anything is imported from `mpi4py`.

So checking whether `h5py` was built with MPI support is difficult, because importing `h5py` submodules will trigger imports from `mpi4py`, and then it's too late (the `MPI_Init` done by `mpi4py` is already done).

I can move this to wherever `mpi4py` is used first, but only once I'm sure where it is imported first, that's not clear to me yet (nor is it clear to me why the tests are failing, but I still need to deep dive into it).

Reply:
I think we only currently import mpi4py in one place (`_set_fapl_mpio`), which you've already found, assuming Cython `cimport` doesn't trigger initialisation.

Reply:
The failure on Azure is also there on master; I'll open a separate issue about that. The Travis failure is in the MPI test run, so that's all yours. :wink:

Reply:
`mpi4py` is also imported from `h5py/h5p.pyx`, so not only in `__set_fapl_mpio`.

I'm not sure that disabling `mpi4py.rc.initialize` can be moved easily elsewhere without re-introducing the problem that the proposed changes are fixing, it should be done as early as possible...

Setting `mpi4py.rc.initialize` to `False` in `__set_fapl_mpio` doesn't make sense, since you actually *want* it to be initialized at that point (that's when you're sure that MPI is actually required).

Reply:
In `h5p.pyx` it's a Cython `cimport` - does even that trigger initialisation? If so, can we suppress it in the same Cython file?

I'm quite keen that a non-MPI build of h5py shouldn't touch mpi4py at all, even if it's there. Maybe @aragilar has some more ideas about how to achieve that.

Reply:
It's not entirely clear to me when exactly the `mpi4py` auto-initialize is triggered.
Certainly not at `import mpi4py`, that's clear (or this would never work).

Maybe @dalcinl can clarify?

Reply:
A bare `import mpi4py` does not trigger `MPI_Init()`, however `from mpi4py import MPI` or `import mpi4py.MPI` does, unless using the `mpi4py.rc.initialize = False` hackery.

That initialize hackery was not meant to be used by libraries, but end users that really need to workaround issues in their main, entry point scripts to inject pre-MPI_Init stuff if really needed.

My take on this is that `from mpi4py import MPI` should not be issued for non-MPI builds of HDF5, and maybe only issueed if the user somehow requests (via imports or calls) the use of MPI-specific features, of couse this is probably not so easy to implement in h5py and there are potential risks of init-deadlocks (talk is cheap, and I'm doing it quite irreponsibly as usual these days).

Reply:
> it's a Cython `cimport` - does even that trigger initialisation?

Cython's `cimport` actually implies a Python `import` behind the scenes.

Reply:
Thanks for the feedback!

It looks like the `MPI_init` that `mpi4py` does by default is then either triggered by the `from mpi4py.MPI import` statement in https://github.com/h5py/h5py/blob/78e665c1991a2a1fbdffa99e3d5f87597881a082/h5py/api_types_ext.pxd#L15
or by the `from mpi4py.mpi_c cimport` statement in https://github.com/h5py/h5py/blob/2660313fcfdb927303d96848014a639b6230abb0/h5py/h5p.pyx#L39

Could the latter imports also trigger the `MPI_Init`?

Reply:
It is be the first one. The second one involves a pxd file that do not expose Python types, just all the pure-C extern declarations corresponding to the MPI API from the usual `mpi.h` include file.

In petsc4py, in order to not impose a hard compile-time dependency on mpi4py (PETSc can be built without MPI), I'm using some hackery that may help h5py:

https://bitbucket.org/petsc/petsc4py/src/54650a1436415513083e5856677a51f594860e80/src/PETSc/petscmpi.pxi#lines-36

Of course, you should be aware of the potential issues of doing imports at the function level.

Reply:
It seems to me that's the proper way forward, to not always blindly do the `from mpi4py.MPI import ...` when `h5py` was built with MPI support enabled.

Disabling the auto-init done by `mpi4py` is biting us in the ass already when using an `h5py` installation that was patched like this, in scripts where `h5py` and `mpi4py` are combined, since whether or not auto-init will happen now depends on the order in which they are imported... 😱 

I'm willing to put more time into this, but I'm not familiar with Cython stuff at all, so this goes beyond what I'm capable of tackling myself in a reasonable amount of time (at least not without significant help from someone more familiar with this).

Is anyone up for helping out, maybe doing a short pair programming session?
Perhaps @aragilar or @takluyver?

Reply:
Thanks @dalcinl for the clarifications.

Would it work to switch the `from mpi4py.MPI cimport ...` in `api_types_ext.pxd` to `from mpi4py.libmpi cimport ...`, like in `h5p.pyx`? If I've understood correctly, the second form lets us make the C types available for Cython but doesn't trigger initialisation.

Reply:
> Would it work to switch the `from mpi4py.MPI cimport ...`

Probably not, because you are *cimporting* the `Comm` class from `mpi4py.MPI`, that is a Python type defined and PyType_Init'ed in `mpi4py.MPI` at import time, and it is not available in `mpi4py.libmpi`, which is just pure extern C MPI stuff.

Folks, if you want to get rid of the compile-time mpi4py dependency (and then maybe runtime dependency if you copy over the `libmpi.pxd` file from mpi ), then you have to:

1. Remove the `cimport` using `mpi4py.MPI`
2. Rewrite `set_fapl_mpio()` something like this:

```
cdef extern from "Python.h":
    ctypedef size_t Py_uintptr_t

def set_fapl_mpio(self, comm, info=None):
    from mpi4py import MPI
    if info is None: info = MPI.INFO_NULL  # info is usually optional!
    assert isinstance(comm, MPI.Intracomm)  # XXX do better!
    assert isinstance(info, MPI.Info)  # XXX do better!
    cdef Py_uintptr_t _comm = MPI._handleof(comm)
    cdef Py_uintptr_t _info = MPI._handleof(info)
    H5Pset_fapl_mpio(self.id, <MPI_Comm>_comm, <MPI_Info>_info)
```
Now mpi4py is no longer a runtime dependency of `h5py`, and it will be imported on-demand. If you do not like the import every time, maybe you can use a private cdef global var initialized to `None` to cache the `MPI` module object.

PS: Be aware that the `_handleof()` utility routine may not be available in older versions of mpi4py previous to 2.0. Do you really need to support that? If yes, Maybe you can still use Cython's conditional compilation with `IF`.

Reply:
Yup, that's roughly the conclusion I was coming to as well - we only use `Comm` and `Info` as argument types for a couple of methods, and we can just use runtime type checking there.

It looks like mpi4py 2.0 came out in October 2015, so I think we can reasonably rely on that, unless there's some specific reason people might prefer to keep using the 1.x series for a long time.

@boegel do you want to have a go at implementing @dalcinl's suggestion? It might be easiest to do it in a new PR, since I think it should make the other changes unnecessary.

Reply:
I can take a stab at it, but like I mentioned I'm not very familiar with Cython.

I'll try and apply the patch proposed by @dalcinl, see if that fixes the runtime issue I'm interested in fixing, and whether the `h5py` tests approve of the change.

It makes sense to tackle that in a new PR indeed, since it's a different approach.

I'll probably have some questions along the way, but I can put those forward in the new PR as they pop up.

Reply:
@boegel - are you still interested in having a go at that? Or would you rather I picked it up>

Reply:
I wouldn't mind looking into it, but I can't make any promises as to when I will get around to it (seems unlikely in the coming days tbh).

If you have a go at it, I'm happy to test what you come up with though, keep me posted!

Reply:
OK, I've had a go in #1552. The MPI tests on CI are passing again, but I haven't tested it beyond that, so having someone who actually uses MPI try it would be great.